server/src/main/java/org/apache/iotdb/db/auth/AuthorizerManager.java

```diff
@@ -342,4 +370,179 @@ public void processLoadSnapshot(File snapshotDir) throws TException, IOException
       snapshotLock.writeLock().unlock();
     }
   }
+
+  public TSStatus checkPath(String username, List<String> allPath, int permission)
+      throws AuthException {
+    snapshotLock.writeLock().lock();
+    try {
+      User user = userCache.getIfPresent(username);
+      boolean status = true;
+      if (user != null) {
+        for (String path : allPath) {
+          if (!user.checkPrivilege(path, permission)) {
+            for (String roleName : user.getRoleList()) {
+              Role role = roleCache.getIfPresent(roleName);
+              // It is detected that the role of the user does not exist in the cache, indicating
+              // that the permission information of the role has changed.
+              // The user cache needs to be initialized
+              if (role == null) {
+                invalidateCache(username, "");
+                status = false;
+                break;
+              }
+              if (!role.checkPrivilege(path, permission)) {
+                status = false;
+              }
```

Comment:
why break when role is null, but not break when role checkPrivilege failed?

Reply:
I've made changes here. You can have a look again